File: python/tasks/crossword-puzzle.py
#!/bin/python

import math
import os
import random
import re
import sys

def check(crossword, word):
    slots = []
    length = len(word)
    for row in range(10):
        for col in range(10):
            validH = True
            validV = True
            for w in range(length):
                if col <= 10 - length:
                    if crossword[row][col + w] not in  ['-', word[w]]:
                        validH = False
                if row <= 10 - length:
                    if crossword[row + w][col] not in  ['-', word[w]]:
                        validV = False
            if validH and col <= 10 - length:
                yield(row, col, 'H')
            if validV and row <= 10 - length:
                yield(row, col, 'V')

def writePuzzle(crossword, word, slot, empty = False):
    row, col, dir = slot;
    for w in range(len(word)):
        letter = empty if empty else word[w]
        if dir == 'H':
            crossword[row][col + w] = letter
        else:
            crossword[row + w][col] = letter

# recombine split crossword_item row lists back into row strings
def merge(crossword):
    board = []
    if isinstance(crossword[0], list):
        for row in crossword:
            board.append(''.join(row))
    else:
        board = crossword
    return board

# Complete the crosswordPuzzle function below.
def crosswordPuzzle(crossword, words):
    if len(words) == 0:
        return crossword
    word = words.pop()
    for slot in check(crossword, word):
        writePuzzle(crossword, word, slot)
        board = crosswordPuzzle(crossword, words)
        if board:
            return merge(board)
        else:
            writePuzzle(crossword, word, slot, '-');
    words.append(word)

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    crossword = []

    for _ in xrange(10):
        crossword_item = list(raw_input())
        # split items into letter list instead of string
        crossword.append(crossword_item)

    # split semi-colon into list
    words = raw_input().split(';')

    result = crosswordPuzzle(crossword, words)
    print result
    fptr.write('\n'.join(result))
    fptr.write('\n')

    fptr.close()

File: js/tasks/crossword-puzzle.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function check(crossword, word) {
    let slots = [];
    for (let row = 0; row < 10; row++) {
        for (let col = 0; col < 10; col++) {
            let validH = true;
            let validV = true;
            for (let w = 0; w < word.length; w++) {
                if (col <= 10 - word.length)
                    if (['-', word[w]].indexOf(crossword[row][col + w]) < 0)
                        validH = false;
                if (row <= 10 - word.length)
                    if (['-', word[w]].indexOf(crossword[row + w][col]) < 0)
                        validV = false;
            }
            
            if (validH && col <= 10 - word.length)
                slots.push([row, col, 'H']);
            if (validV && row <= 10 - word.length)
                slots.push([row, col, 'V']);
        }
    }
    return slots;
}

function writePuzzle(crossword, word, slot, empty = false) {
    let [row, col, dir] = slot;
    for (let w = 0; w < word.length; w++) {
        let letter = empty ? empty : word[w];
        if (dir === 'H') {
            crossword[row][col + w] = letter;
        } else {
            crossword[row + w][col] = letter;
        }
    }
}

// recombine split crosswordItem row arrays back into row strings
function merge(crossword) {
    let board = [];
    if (Array.isArray(crossword[0])) {
        for (let row = 0; row < crossword.length; row++) {
            board.push(crossword[row].join(''));
        }
    } else {
        board = crossword;
    }
    return board;
}

// Complete the crosswordPuzzle function below.
function crosswordPuzzle(crossword, hints) {
    if (hints.length === 0) return crossword;
    let word = hints.pop();
    let slot = check(crossword, word);
    // check(crossword, word).forEach((slot) => did not work, using forloop
    for (let s = 0; s < slot.length; s++) {
        writePuzzle(crossword, word, slot[s]);
        let board = crosswordPuzzle(crossword, hints);
        if (board) {
            return merge(board);
        } else {
            writePuzzle(crossword, word, slot[s], '-');        
        }
    }
    hints.push(word);
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    let crossword = [];

    for (let i = 0; i < 10; i++) {
        const crosswordItem = readLine().split('');
        // string => read only letters, so split row into col array before push
        crossword.push(crosswordItem);
    }

    const words = readLine().split(';');

    const result = crosswordPuzzle(crossword, words);
    ws.write(result.join('\n') + '\n');

    ws.end();
}

